### ISSUE TITLE
Match scales (PCA) via Command Line Interface

### ISSUE THREAD

**Issue by raphael2692** [CloudCompare/CloudCompare#1034]
Hi all, 

I am trying to use Cloud Compare via Command Line to automate the alignement of 2 clouds (obtained from sampling 2 meshes). 

With the GUI I am able to achieve this with these operations:

1. Tools -> Registration -> Match bounding box centers
2. Tools -> Registration  -> Match scales (PCA)
3. Tools -> Registration -> ICP Fine registration

I am working on replicate these operations via CLI but it seems the _Match scales_ filter is missing. Is it really so? Can you suggest me a work around? Thanks!

**Comment by dgirardeau** [comment]
Indeed, the 'match scales' feature has probably not been ported to the command line yet ;)

Have you tried ICP with variable scale instead? (-ADJUST_SCALE)

**Comment by raphael2692** [comment]
Hi @dgirardeau , thanks for your response. 

I tried ICP with that parameter only but the result is either not homomorphic (aka the the aligning cloud is deformed) or completely wrong (the cloud is caled down by a huge factor and fitted a small patch of the template). 

My ultimate goal is to use a template cloud to automatically scale different clouds of the same scene created via photogrammetry pipelines (structure-from-motion + multiview stereo, like openMVG). 

Do you have any other suggestion? 😊

**Comment by dgirardeau** [comment]
Apart from instrumenting the command line mode to make this method available (you or someone else), no ;)

**Comment by raphael2692** [comment]
Hi @dgirardeau , can you help me understanding where this function is located in the source code?

Also, it would be interesting to implement a scale range costraint, as it would dramatically reduce local minima pitfalls (resulting in small patch fitting) as described in [this paper](https://www.researchgate.net/publication/242506513_Scaling_iterative_closest_point_algorithm_for_registration_of_mD_point_sets).  Thanks!

**Comment by dgirardeau** [comment]
The method itself is in ccLibAlgorithms::ApplyScaleMatchingAlgorithm (https://github.com/CloudCompare/CloudCompare/blob/8f4cc8abfb697a820f32feef02a41ea2c207623c/qCC/ccLibAlgorithms.cpp#L575). You can see how it is called in the mainwindow.cpp file: https://github.com/CloudCompare/CloudCompare/blob/42203ddcda49a169b10cfe8b6d340548f53ff582/qCC/mainwindow.cpp#L5233

And I believe it should be easy to add such min/max range constraints, especially for the command line version (for the GUI version, you would have to modify the tool dialog - that's always possible of course, but it will requires some knowledge of how Qt works). If the tool is modified, I can handle the modification of the dialog.

**Comment by somtri** [comment]
Both requests in this thread are in the command line now:

- Match scales: `-MATCH_SCALES {BB_MAX_DIM|BB_VOLUME|PCA_MAX_DIM|ICP} [-REFERENCE {index}] [-RMS_DIFF {value}] [-OVERLAP {percent}]`. `PCA_MAX_DIM` is the PCA method from the GUI dialog.
- Rotation and translation constraints for ICP: `-ICP -ROT {XYZ|X|Y|Z|NONE}`, plus `-SKIP_TX`, `-SKIP_TY` and `-SKIP_TZ`.

@dgirardeau on the min/max scale range you suggested here: before writing it, what should happen when the computed scale falls outside the range? Clamping to the nearest bound, rejecting the match and leaving the cloud unscaled, and warning while continuing are all defensible, and they give different results.

On the plumbing, `ccLibAlgorithms::ApplyScaleMatchingAlgorithm` already ends in defaulted arguments and has two callers, the command line one and `mainwindow.cpp`. Adding min and max as defaulted parameters would leave the GUI caller untouched.

**Comment by dgirardeau** [comment]
I think the idea was to constrain the scales and clamp the values indeed. This requires to change CCCoreLib as well probably?

**Comment by somtri** [comment]
No CCCoreLib change needed for the clamping itself. All four methods end up as a plain `double` inside `ApplyScaleMatchingAlgorithm`, and the factor that actually gets applied is computed there:

```cpp
double scaled = algo == ICP_SCALE ? scales[i] : scales[refEntityIndex] / scales[i];
```

Clamping that one value covers all four methods, so the diff stays inside qCC.

CCCoreLib would come into it for the other reading of the request: constraining the scale during the ICP search itself, which is how I read the paper linked earlier. `ICPRegistrationTools::Parameters` has no scale bounds today, so bounding the search would mean adding them there. Clamping afterwards bounds the result but does not steer ICP away from a local minimum, so it is a guardrail rather than a fix for the small-patch fitting described earlier in this thread.

I have implemented the clamping version you described: a factor outside the range is clamped to the nearest bound, and the warning names both the original factor and the bound applied. The two new sub-options `-MIN_SCALE` and `-MAX_SCALE` are optional, with no limit by default. Both are defaulted parameters on `ApplyScaleMatchingAlgorithm`, so the `mainwindow.cpp` caller is untouched and the dialog side stays yours.

I can open a PR with it if you want it.
